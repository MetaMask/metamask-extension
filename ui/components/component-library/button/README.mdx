import { Story, Canvas, ArgsTable } from '@storybook/addon-docs';
import { Button } from './button';
import { ButtonBase } from '../button-base';
import { ButtonLink } from '../button-link';

# Button (deprecated use @metamask/design-system-react)

The `Button` is used for user actions. It unifies `ButtonPrimary`, `ButtonSecondary` and `ButtonLink`.

[MetaMask Design System Guides](https://www.notion.so/MetaMask-Design-System-Guides-Design-f86ecc914d6b4eb6873a122b83c12940)

<Canvas>
  <Story id="components-componentlibrary-button--default-story" />
</Canvas>

## Props

<ArgsTable of={Button} />

### Variant

Use the `variant` prop and the `ButtonVariant` enum from `./ui/components/component-library` to change the `Button` variant.

Possible variants include:

- `ButtonVariant.Primary` (default)
- `ButtonVariant.Secondary`
- `ButtonVariant.Link`

<Canvas>
  <Story id="components-componentlibrary-button--variant" />
</Canvas>

```jsx
import { Button, ButtonVariant } from '../ui/component-library/button';

<Button variant={ButtonVariant.Primary}>Button Primary</Button>
<Button variant={ButtonVariant.Secondary}>Button Secondary</Button>
<Button variant={ButtonVariant.Link}>Button Link</Button>
```

### Size

Use the `size` prop and the `ButtonSize` enum from `./ui/components/component-library` to change the size of `Button`. Defaults to `Buttonsize.Md`

Possible sizes include:

- `Buttonsize.Inherit` inherits the font-size of the parent element. For [ButtonLink](/docs/components-componentlibrary-buttonlink--default-story) uses only
- `Buttonsize.Auto` inherits auto height, but keeps the same button font-size. For [ButtonLink](/docs/components-componentlibrary-buttonlink--default-story) uses only
- `Buttonsize.Sm` 32px
- `Buttonsize.Md` 40px
- `Buttonsize.Lg` 48px

<Canvas>
  <Story id="components-componentlibrary-button--size-story" />
</Canvas>

```jsx
import { Button, ButtonSize } from '../../component-library';

<Button size={Buttonsize.Inherit} />
<Button size={Buttonsize.Sm} />
<Button size={Buttonsize.Md} />
<Button size={Buttonsize.Lg} />
```

### Danger

Use the `danger` boolean prop to change the `Button` to danger color.

<Canvas>
  <Story id="components-componentlibrary-button--danger" />
</Canvas>

```jsx
import { Button } from '../../component-library';

<Button>Normal</Button>
<Button danger>Danger</Button>
```

### Href

When an `href` prop is passed it will change the element to an anchor(`a`) tag.

<Canvas>
  <Story id="components-componentlibrary-button--href" />
</Canvas>

```jsx
import { Button } from '../../component-library';

<Button href="/">Anchor Element</Button>;
```

### Block

Use boolean `block` prop to quickly enable a full width block button

<Canvas>
  <Story id="components-componentlibrary-button--block" />
</Canvas>

```jsx
import { Button } from '../../component-library';

<Button>Default Button</Button>
<Button block>Block Button</Button>
```

### As

Use the `as` box prop to change the element of `Button`. Defaults to `button`.

When an `href` prop is passed it will change the element to an anchor(`a`) tag.

Button `as` options:

- `button`
- `a`

<Canvas>
  <Story id="components-componentlibrary-button--as" />
</Canvas>

```jsx
import { Button } from '../../component-library';


<Button as="button">Button Element</Button>
<Button as="a" href="#">
  Anchor Element
</Button>
```

### Disabled

Use the boolean `disabled` prop to disable button

<Canvas>
  <Story id="components-componentlibrary-button--disabled" />
</Canvas>

```jsx
import { Button } from '../../component-library';

<Button disabled>Disabled Button</Button>;
```

### Loading

Use the boolean `loading` prop to set loading spinner

<Canvas>
  <Story id="components-componentlibrary-button--loading" />
</Canvas>

```jsx
import { Button } from '../../component-library';

<Button loading>Loading Button</Button>;
```

### Icon Name

Use the `startIconName` and/or `endIconName` prop and the `IconName` enum from `./ui/components/component-library/icon` to select icon.

Use the [IconSearch](/story/components-componentlibrary-icon--default-story) story to find the icon you want to use.

<Canvas>
  <Story id="components-componentlibrary-button--start-icon-name" />
</Canvas>

```jsx
import { Button, IconName } from '../../component-library';

<Button startIconName={IconName.AddSquare}>Button</Button>;
```

<Canvas>
  <Story id="components-componentlibrary-button--end-icon-name" />
</Canvas>

```jsx
import { Button, IconName } from '../../component-library';

<Button endIconName={IconName.Arrow2Right}>Button</Button>;
```
