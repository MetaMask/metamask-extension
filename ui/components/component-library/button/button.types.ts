import type { PolymorphicComponentPropWithRef } from '../box';
import type { ButtonPrimaryStyleUtilityProps } from '../button-primary/button-primary.types';
import type { ButtonSecondaryStyleUtilityProps } from '../button-secondary/button-secondary.types';
import type { ButtonLinkStyleUtilityProps } from '../button-link/button-link.types';

export enum ButtonSize {
  Sm = 'sm',
  Md = 'md',
  Lg = 'lg',
  Inherit = 'inherit',
  Auto = 'auto',
}

export enum ButtonVariant {
  Primary = 'primary',
  Secondary = 'secondary',
  Link = 'link',
}

type ValidButtonSize = ButtonSize.Sm | ButtonSize.Md | ButtonSize.Lg;

type ButtonPropsByVariant = {
  [ButtonVariant.Primary]: {
    variant?: ButtonVariant.Primary;
    size?: ValidButtonSize; // Allows for only ButtonSize.Sm, ButtonSize.Md, ButtonSize.Lg
  } & Omit<ButtonPrimaryStyleUtilityProps, 'size' | 'variant'>;
  [ButtonVariant.Secondary]: {
    variant?: ButtonVariant.Secondary;
    size?: ValidButtonSize; // Allows for only ButtonSize.Sm, ButtonSize.Md, ButtonSize.Lg
  } & Omit<ButtonSecondaryStyleUtilityProps, 'size' | 'variant'>;
  [ButtonVariant.Link]: {
    variant?: ButtonVariant.Link;
    size?: ButtonSize;
  } & Omit<ButtonLinkStyleUtilityProps, 'size' | 'variant'>;
};

type ButtonPropsMap = {
  [variant in ButtonVariant]: ButtonPropsByVariant[variant];
};

export type ButtonProps<C extends React.ElementType> =
  PolymorphicComponentPropWithRef<C, ButtonPropsMap[ButtonVariant]>;

export type ButtonComponent = <C extends React.ElementType = 'button' | 'a'>(
  props: ButtonProps<C>,
) => React.ReactElement | null;
