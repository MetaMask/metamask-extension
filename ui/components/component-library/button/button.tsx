import React from 'react';
import type { PolymorphicRef } from '../box';
import { ButtonPrimary } from '../button-primary';
import { ButtonSecondary } from '../button-secondary';
import { ButtonLink } from '../button-link';
import type { ButtonPrimaryProps } from '../button-primary/button-primary.types';
import type { ButtonSecondaryProps } from '../button-secondary/button-secondary.types';
import type { ButtonLinkProps } from '../button-link/button-link.types';
import type { ButtonProps, ButtonComponent } from './button.types';

import { ButtonVariant } from './button.types';

export const Button: ButtonComponent = React.forwardRef(
  <C extends React.ElementType = 'button' | 'a'>(
    { variant, ...props }: ButtonProps<C>,
    ref?: PolymorphicRef<C>,
  ) => {
    switch (variant) {
      case ButtonVariant.Primary:
        return (
          <ButtonPrimary ref={ref} {...(props as ButtonPrimaryProps<C>)} />
        );
      case ButtonVariant.Secondary:
        return (
          <ButtonSecondary ref={ref} {...(props as ButtonSecondaryProps<C>)} />
        );
      case ButtonVariant.Link:
        return <ButtonLink ref={ref} {...(props as ButtonLinkProps<C>)} />;
      default:
        return (
          <ButtonPrimary ref={ref} {...(props as ButtonPrimaryProps<C>)} />
        );
    }
  },
);
