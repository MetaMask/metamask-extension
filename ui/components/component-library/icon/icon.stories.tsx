import React, { useState } from 'react';
import { StoryFn, Meta } from '@storybook/react';

import {
  AlignItems,
  Display,
  IconColor,
  FlexDirection,
  JustifyContent,
  TextVariant,
  FlexWrap,
  TextAlign,
  BackgroundColor,
  BorderColor,
  Color,
  TextColor,
  BorderRadius,
} from '../../../helpers/constants/design-system';

import { Text } from '../text';
import { Icon } from './icon';
import { IconName, IconSize } from './icon.types';

import README from './README.mdx';
import { Box } from '../box';
import { Label } from '../label';
import { TextFieldSearch } from '../text-field-search';
import { TextField, TextFieldSize } from '../text-field';
import { ButtonIcon, ButtonIconSize } from '../button-icon';
import { ButtonLink, ButtonLinkSize } from '../button-link';

export default {
  title: 'Components/ComponentLibrary/Icon',
  component: Icon,
  parameters: {
    docs: {
      page: README,
    },
  },
  args: {
    name: IconName.AddSquare,
    color: IconColor.inherit,
    size: IconSize.Md,
  },
} as Meta<typeof Icon>;

export const DefaultStory: StoryFn<typeof Icon> = (args) => {
  const [search, setSearch] = useState('');
  const iconList = Object.keys(IconName)
    .filter(
      (item) =>
        search === '' ||
        item.toLowerCase().includes(search.toLowerCase().replace(' ', '_')),
    )
    .sort();

  const handleSearch = (e) => {
    setSearch(e.target.value);
  };

  const handleOnClear = () => {
    setSearch('');
  };

  return (
    <>
      <Text as="h2" marginBottom={2} variant={TextVariant.headingMd}>
        Icon search
      </Text>
      <Box
        display={Display.Grid}
        gap={2}
        style={{
          gridTemplateColumns: 'repeat(auto-fill, minmax(180px, 1fr))',
        }}
      >
        <Box
          style={{ gridColumnStart: 1, gridColumnEnd: 3 }}
          display={Display.Flex}
          flexDirection={FlexDirection.Column}
        >
          {/* TODO replace with FormTextField */}
          <Label htmlFor="icon-search">IconName</Label>
          <TextFieldSearch
            id="icon-search"
            marginBottom={4}
            onChange={handleSearch}
            clearButtonOnClick={handleOnClear}
            value={search}
            placeholder="Search icon name"
          />
        </Box>
      </Box>
      {iconList.length > 0 ? (
        <Box
          display={Display.Grid}
          gap={2}
          style={{
            gridTemplateColumns: 'repeat(auto-fill, minmax(180px, 1fr))',
          }}
        >
          {iconList.map((item) => {
            return (
              <Box
                borderColor={BorderColor.borderMuted}
                borderRadius={BorderRadius.MD}
                display={Display.Flex}
                flexDirection={FlexDirection.Column}
                alignItems={AlignItems.center}
                justifyContent={JustifyContent.center}
                padding={4}
                key={item}
              >
                <Icon marginBottom={2} {...args} name={IconName[item]} />
                <TextField
                  placeholder={item}
                  value={item}
                  readOnly
                  size={TextFieldSize.Sm}
                  inputProps={{
                    variant: TextVariant.bodyXs,
                    textAlign: TextAlign.Center,
                  }}
                  backgroundColor={BackgroundColor.backgroundAlternative}
                  endAccessory={
                    <ButtonIcon
                      iconName={IconName.Copy}
                      size={ButtonIconSize.Sm}
                      color={IconColor.iconAlternative}
                      ariaLabel="Copy to clipboard"
                      title="Copy to clipboard"
                      onClick={() => {
                        const tempEl = document.createElement('textarea');
                        tempEl.value = item;
                        document.body.appendChild(tempEl);
                        tempEl.select();
                        document.execCommand('copy');
                        document.body.removeChild(tempEl);
                      }}
                    />
                  }
                />
              </Box>
            );
          })}
        </Box>
      ) : (
        <Text>
          No matches. Please try again or ask in the{' '}
          <ButtonLink
            size={ButtonLinkSize.Inherit}
            color={TextColor.primaryDefault}
            href="https://consensys.slack.com/archives/C0354T27M5M"
            target="_blank"
          >
            #metamask-design-system
          </ButtonLink>{' '}
          channel on slack.
        </Text>
      )}
    </>
  );
};
DefaultStory.storyName = 'Default';

export const Name: StoryFn<typeof Icon> = (args) => (
  <>
    <Box display={Display.Flex} flexWrap={FlexWrap.Wrap} gap={2}>
      {Object.keys(IconName).map((item) => {
        return (
          <Box
            borderColor={BorderColor.borderMuted}
            borderRadius={BorderRadius.MD}
            display={Display.Flex}
            flexDirection={FlexDirection.Column}
            alignItems={AlignItems.center}
            justifyContent={JustifyContent.center}
            padding={4}
            key={item}
          >
            <Icon {...args} name={IconName[item]} />
          </Box>
        );
      })}
    </Box>
  </>
);

export const SizeStory: StoryFn<typeof Icon> = (args) => (
  <>
    <Box
      display={Display.Flex}
      alignItems={AlignItems.baseline}
      gap={1}
      marginBottom={4}
    >
      <Icon {...args} size={IconSize.Xs} />
      <Icon {...args} size={IconSize.Sm} />
      <Icon {...args} size={IconSize.Md} />
      <Icon {...args} size={IconSize.Lg} />
      <Icon {...args} size={IconSize.Xl} />
    </Box>
    <Text variant={TextVariant.displayMd}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.headingLg}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.headingMd}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.headingSm}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.bodyLgMedium}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.bodyMd}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.bodySm}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
    <Text variant={TextVariant.bodyXs}>
      inherits the font-size of the parent element.{' '}
      <Icon {...args} size={IconSize.Inherit} />
    </Text>
  </>
);
SizeStory.storyName = 'Size';

export const ColorStory: StoryFn<typeof Icon> = (args) => (
  <Box display={Display.Flex} alignItems={AlignItems.baseline}>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.inherit} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.iconDefault} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.iconAlternative} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.iconMuted} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.overlayDefault}
    >
      <Icon {...args} color={IconColor.overlayInverse} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.primaryDefault} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.primaryDefault}
    >
      <Icon {...args} color={IconColor.primaryInverse} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.errorDefault} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.errorDefault}
    >
      <Icon {...args} color={IconColor.errorInverse} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.successDefault} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.successDefault}
    >
      <Icon {...args} color={IconColor.successInverse} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.warningDefault} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.warningDefault}
    >
      <Icon {...args} color={IconColor.warningInverse} />
    </Box>
    <Box padding={1} display={Display.Flex} alignItems={AlignItems.center}>
      <Icon {...args} color={IconColor.infoDefault} />
    </Box>
    <Box
      padding={1}
      display={Display.Flex}
      alignItems={AlignItems.center}
      backgroundColor={BackgroundColor.infoDefault}
    >
      <Icon {...args} color={IconColor.infoInverse} />
    </Box>
  </Box>
);
ColorStory.storyName = 'Color';

export const LayoutAndSpacing = () => (
  <Box display={Display.Flex} flexDirection={FlexDirection.Column} gap={4}>
    <Box display={Display.Flex} alignItems={AlignItems.center}>
      <Icon
        name={IconName.Check}
        color={IconColor.successDefault}
        marginRight={1}
      />
      <Text>Always allow you to opt-out via Settings</Text>
    </Box>
    <Box
      as="button"
      display={Display.Flex}
      alignItems={AlignItems.center}
      borderRadius={BorderRadius.pill}
      backgroundColor={BackgroundColor.primaryMuted}
      marginRight="auto"
    >
      <Text color={Color.primaryDefault}>
        0x79fAaFe7B6D5DB5D8c63FE88DFF0AF1Fe53358db
      </Text>
      <Icon
        name={IconName.Copy}
        color={IconColor.primaryDefault}
        marginLeft={1}
      />
    </Box>
    <Box
      as="button"
      display={Display.Flex}
      alignItems={AlignItems.center}
      padding={4}
      borderColor={BorderColor.borderMuted}
      backgroundColor={BackgroundColor.backgroundDefault}
    >
      <Icon name={IconName.Add} color={IconColor.iconDefault} marginRight={2} />
      <Text>Create account</Text>
    </Box>
    <Label>
      Custom spending cap{' '}
      <Icon name={IconName.Info} size={IconSize.Inherit} marginLeft={1} />
    </Label>
    <div>
      <Text>
        <Icon
          name={IconName.Warning}
          size={IconSize.Inherit}
          marginLeft={1}
          color={IconColor.warningDefault}
        />{' '}
        Warning
      </Text>
    </div>
  </Box>
);
