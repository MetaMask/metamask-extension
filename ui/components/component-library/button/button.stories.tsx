import React from 'react';
import { StoryFn, Meta } from '@storybook/react';
import {
  AlignItems,
  Display,
  FlexDirection,
  TextVariant,
} from '../../../helpers/constants/design-system';
import { Text } from '../text';
import README from './README.mdx';
import { Button } from './button';
import { IconName } from '../icon';
import { ButtonSize, ButtonVariant } from './button.types';
import { Box } from '../box';

export default {
  title: 'Components/ComponentLibrary/Button',
  component: Button,
  parameters: {
    docs: {
      page: README,
    },
    controls: { sort: 'alpha' },
  },
  argTypes: {
    as: {
      control: 'select',
      options: ['button', 'a'],
    },
    block: {
      control: 'boolean',
    },
    children: {
      control: 'text',
    },
    className: {
      control: 'text',
    },
    danger: {
      control: 'boolean',
    },
    disabled: {
      control: 'boolean',
    },
    href: {
      control: 'text',
    },
    startIconName: {
      control: 'select',
      options: Object.values(IconName),
    },
    endIconName: {
      control: 'select',
      options: Object.values(IconName),
    },
    startIconProps: {
      control: 'object',
    },
    endIconProps: {
      control: 'object',
    },
    loading: {
      control: 'boolean',
    },
    size: {
      control: 'select',
      options: Object.values(ButtonSize),
    },
    variant: {
      options: Object.values(ButtonVariant),
      control: 'select',
    },
  },
  args: {
    children: 'Button',
  },
} as Meta<typeof Button>;

const Template: StoryFn<typeof Button> = (args) => <Button {...args} />;

export const DefaultStory = Template.bind({});
DefaultStory.storyName = 'Default';

export const Variant: StoryFn<typeof Button> = () => (
  <Box display={Display.Flex} gap={1}>
    <Button variant={ButtonVariant.Primary}>Button Primary</Button>
    <Button variant={ButtonVariant.Secondary}>Button Secondary</Button>
    <Button variant={ButtonVariant.Link}>Button Link</Button>
  </Box>
);

export const SizeStory: StoryFn<typeof Button> = (args) => (
  <>
    <Box
      display={Display.Flex}
      alignItems={AlignItems.baseline}
      gap={1}
      marginBottom={3}
    >
      <Button {...args} variant={ButtonVariant.Primary} size={ButtonSize.Sm}>
        Small Button
      </Button>
      <Button {...args} size={ButtonSize.Md}>
        Medium (Default) Button
      </Button>
      <Button {...args} size={ButtonSize.Lg}>
        Large Button
      </Button>
      <Button {...args} variant={ButtonVariant.Link}>
        Auto ButtonLink
      </Button>
    </Box>
    <Text variant={TextVariant.bodySm}>
      <Button {...args} variant={ButtonVariant.Link} size={ButtonSize.Inherit}>
        Button Inherit
      </Button>{' '}
      inherits the font-size of the parent element. Inherit size only used for
      ButtonLink.
    </Text>
  </>
);
SizeStory.storyName = 'Size';

export const Danger: StoryFn<typeof Button> = (args) => (
  <Box display={Display.Flex} gap={1}>
    <Button {...args}>Normal</Button>
    {/* Test Anchor tag to match exactly as button */}
    <Button as="a" {...args} href="#" danger>
      Danger
    </Button>
  </Box>
);

export const Href: StoryFn<typeof Button> = (args) => (
  <Button {...args}>Anchor Element</Button>
);

Href.args = {
  href: '/metamask',
};

export const Block: StoryFn<typeof Button> = (args) => (
  <>
    <Button {...args} marginBottom={2}>
      Default Button
    </Button>
    <Button {...args} block marginBottom={2}>
      Block Button
    </Button>
  </>
);

export const As: StoryFn<typeof Button> = (args) => (
  <Box display={Display.Flex} flexDirection={FlexDirection.Row} gap={2}>
    <Button {...args}>Button Element</Button>
    <Button as="a" href="#" {...args}>
      Anchor Element
    </Button>
  </Box>
);

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Loading = Template.bind({});
Loading.args = {
  loading: true,
};

export const StartIconName = Template.bind({});
StartIconName.args = {
  startIconName: IconName.AddSquare,
};

export const EndIconName = Template.bind({});
EndIconName.args = {
  endIconName: IconName.AddSquare,
};
