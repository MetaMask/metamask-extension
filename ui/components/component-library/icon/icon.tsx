import React from 'react';
import classnames from 'classnames';

import { IconColor, Display } from '../../../helpers/constants/design-system';

import { Box, BoxProps } from '../box';
import type { PolymorphicRef } from '../box';

import { IconSize, IconProps, IconComponent } from './icon.types';

export const Icon: IconComponent = React.forwardRef(
  <C extends React.ElementType = 'span'>(
    {
      name,
      size = IconSize.Md,
      color = IconColor.inherit,
      className = '',
      style,
      ...props
    }: IconProps<C>,
    ref?: PolymorphicRef<C>,
  ) => (
    <Box
      className={classnames(className, 'mm-icon', `mm-icon--size-${size}`)}
      ref={ref}
      as="span"
      display={Display.InlineBlock}
      color={color}
      style={{
        /**
         * To reduce the possibility of injection attacks
         * the icon component uses mask-image instead of rendering
         * the svg directly.
         */
        maskImage: `url('./images/icons/${String(name)}.svg')`,
        WebkitMaskImage: `url('./images/icons/${String(name)}.svg')`,
        ...style,
      }}
      {...(props as BoxProps<C>)}
    />
  ),
);
