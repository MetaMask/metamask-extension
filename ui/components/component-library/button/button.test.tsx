/* eslint-disable jest/require-top-level-describe */
import { render } from '@testing-library/react';
import React from 'react';
import { IconName } from '../icon';
import { Button } from './button';
import { ButtonSize, ButtonVariant } from './button.types';

describe('Button', () => {
  it('should render button element correctly', () => {
    const { getByTestId, getByText, container } = render(
      <Button data-testid="button">Button</Button>,
    );
    expect(getByText('Button')).toBeDefined();
    expect(container.querySelector('button')).toBeDefined();
    expect(getByTestId('button')).toHaveClass('mm-button-base');
    expect(container).toMatchSnapshot();
  });

  it('should render anchor element correctly', () => {
    const { getByTestId, container } = render(
      <Button as="a" data-testid="button">
        Button
      </Button>,
    );
    expect(getByTestId('button')).toHaveClass('mm-button-base');
    const anchor = container.getElementsByTagName('a').length;
    expect(anchor).toBe(1);
  });

  it('should render anchor element correctly by href only being passed', () => {
    const { getByTestId, container } = render(
      <Button href="/metamask" data-testid="button">
        Visit Site
      </Button>,
    );
    expect(getByTestId('button')).toHaveClass('mm-button-base');
    const anchor = container.getElementsByTagName('a').length;
    expect(anchor).toBe(1);
  });

  it('should render button as block', () => {
    const { getByTestId } = render(<Button block data-testid="block" />);
    expect(getByTestId('block')).toHaveClass(`mm-button-base--block`);
  });

  it('should render with different button types', () => {
    const { getByTestId, container } = render(
      <>
        <Button
          variant={ButtonVariant.Primary}
          data-testid={ButtonVariant.Primary}
        >
          Button
        </Button>
        <Button
          variant={ButtonVariant.Secondary}
          data-testid={ButtonVariant.Secondary}
        >
          Button
        </Button>
        <Button variant={ButtonVariant.Link} data-testid={ButtonVariant.Link}>
          Button
        </Button>
      </>,
    );
    expect(getByTestId(ButtonVariant.Primary)).toHaveClass(
      `mm-button-${ButtonVariant.Primary}`,
    );
    expect(getByTestId(ButtonVariant.Secondary)).toHaveClass(
      `mm-button-${ButtonVariant.Secondary}`,
    );
    expect(getByTestId(ButtonVariant.Link)).toHaveClass(
      `mm-button-${ButtonVariant.Link}`,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render with different size classes', () => {
    const { getByTestId } = render(
      <>
        <Button
          variant={ButtonVariant.Link}
          size={ButtonSize.Inherit}
          data-testid={ButtonSize.Inherit}
        >
          Button {ButtonSize.Inherit}
        </Button>
        <Button size={ButtonSize.Sm} data-testid={ButtonSize.Sm}>
          Button {ButtonSize.Sm}
        </Button>
        <Button size={ButtonSize.Md} data-testid={ButtonSize.Md}>
          Button {ButtonSize.Md}
        </Button>
        <Button size={ButtonSize.Lg} data-testid={ButtonSize.Lg}>
          Button {ButtonSize.Lg}
        </Button>
      </>,
    );
    expect(getByTestId(ButtonSize.Inherit)).toHaveClass(
      `mm-button-link--size-${ButtonSize.Inherit}`,
    );
    expect(getByTestId(ButtonSize.Sm)).toHaveClass(
      `mm-button-base--size-${ButtonSize.Sm}`,
    );
    expect(getByTestId(ButtonSize.Md)).toHaveClass(
      `mm-button-base--size-${ButtonSize.Md}`,
    );
    expect(getByTestId(ButtonSize.Lg)).toHaveClass(
      `mm-button-base--size-${ButtonSize.Lg}`,
    );
  });

  it('should render with added classname', () => {
    const { getByTestId } = render(
      <Button data-testid="classname" className="mm-button-base--test">
        Button
      </Button>,
    );
    expect(getByTestId('classname')).toHaveClass('mm-button-base--test');
  });

  it('should render with different button states', () => {
    const { getByTestId } = render(
      <>
        <Button loading data-testid="loading">
          Button
        </Button>
        <Button disabled data-testid="disabled">
          Button
        </Button>
      </>,
    );
    expect(getByTestId('loading')).toHaveClass(`mm-button-base--loading`);
    expect(getByTestId('disabled')).toHaveClass(`mm-button-base--disabled`);
  });
  it('should render with icon', () => {
    const { getByTestId } = render(
      <Button
        data-testid="icon"
        startIconName={IconName.AddSquare}
        startIconProps={{
          'data-testid': 'start-button-icon',
          name: IconName.AddSquare,
        }}
      >
        Button
      </Button>,
    );

    expect(getByTestId('start-button-icon')).toBeDefined();
  });
});

it('should render as danger', () => {
  const { getByTestId } = render(
    <>
      <Button danger data-testid="danger">
        Button Danger
      </Button>
    </>,
  );

  expect(getByTestId('danger')).toHaveClass('mm-button-primary--type-danger');
});
