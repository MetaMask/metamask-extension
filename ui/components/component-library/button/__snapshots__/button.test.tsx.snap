// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button element correctly 1`] = `
<div>
  <button
    class="mm-box mm-text mm-button-base mm-button-base--size-md mm-button-primary mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
    data-testid="button"
    data-theme="light"
  >
    Button
  </button>
</div>
`;

exports[`Button should render with different button types 1`] = `
<div>
  <button
    class="mm-box mm-text mm-button-base mm-button-base--size-md mm-button-primary mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
    data-testid="primary"
    data-theme="light"
  >
    Button
  </button>
  <button
    class="mm-box mm-text mm-button-base mm-button-base--size-md mm-button-secondary mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent mm-box--rounded-pill mm-box--border-color-primary-default box--border-style-solid box--border-width-1"
    data-testid="secondary"
  >
    Button
  </button>
  <button
    class="mm-box mm-text mm-button-base mm-button-link mm-button-link--size-auto mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
    data-testid="link"
  >
    Button
  </button>
</div>
`;
